File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressQuantumTreeList/Demos/CBuilder/ScrollbarAnnotationsDemo/ScrollbarAnnotationsDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ScrollbarAnnotationsDemoMain.h"
#include "CustomAnnotationSettings.h"
#include "CarsData.h"
#include "ScrollbarAnnotationsDemoData.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxDBTL"
#pragma link "cxGraphics"
#pragma link "cxInplaceContainer"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxMaskEdit"
#pragma link "cxStyles"
#pragma link "cxTextEdit"
#pragma link "cxTL"
#pragma link "cxTLData"
#pragma link "cxCheckBox"
#pragma resource "*.dfm"
TScrollbarAnnotationsDemoMainForm *ScrollbarAnnotationsDemoMainForm;
String SCustomAnnotationStrs[3];
//---------------------------------------------------------------------------
__fastcall TScrollbarAnnotationsDemoMainForm::TScrollbarAnnotationsDemoMainForm(TComponent* Owner)
	: TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actScrollAnnotationsActiveExecute(TObject *Sender)

{
  cxDBTreeList1->ScrollbarAnnotations->Active = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
Variant TScrollbarAnnotationsDemoMainForm::GetValue(int ARecordIndex, TcxTreeListColumn *AColumn)
{
  return (cxDBTreeList1->AbsoluteVisibleItems[ARecordIndex]->Values[AColumn->ItemIndex]);
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actShowErrorsExecute(TObject *Sender)

{
  cxDBTreeList1->ScrollbarAnnotations->ShowErrors = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actShowSearchResultsExecute(TObject *Sender)

{
  cxDBTreeList1->ScrollbarAnnotations->ShowSearchResults = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actShowFocusedRowExecute(TObject *Sender)

{
  cxDBTreeList1->ScrollbarAnnotations->ShowFocusedRow = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actShowSelectedRowsExecute(TObject *Sender)

{
  cxDBTreeList1->ScrollbarAnnotations->ShowSelectedRows = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actCustomAnnotationSettingsExecute(TObject *Sender)

{
  if (!frmCustomAnnotationSettings)
  {
	Application->CreateForm(__classid(TfrmCustomAnnotationSettings), &frmCustomAnnotationSettings);
	frmCustomAnnotationSettings->Initialize(cxDBTreeList1->ScrollbarAnnotations->CustomAnnotations);
  };
  frmCustomAnnotationSettings->Show();
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::actVacancyExecute(TObject *Sender)

{
  cxDBTreeList1->ScrollbarAnnotations->CustomAnnotations->Items[dynamic_cast<TAction *>(Sender)->Tag]->Visible = dynamic_cast<TAction *>(Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TScrollbarAnnotationsDemoMainForm::cxDBTreeList1PopulateCustomScrollbarAnnotationRowIndexList(TObject *Sender,
          int AAnnotationIndex, TdxScrollbarAnnotationRowIndexList *ARowIndexList)
{
  bool ACondition = false;
  String s;
  Variant AValue;
  for (int i = 0; i < cxDBTreeList1->AbsoluteVisibleCount; i++)
  {
	switch (AAnnotationIndex)
	{
	  case 0:
		  AValue = GetValue(i, clVACANCY);
		  ACondition = !VarIsNull(AValue) && (AValue == VARIANT_TRUE);
		  break;
	  case 1:
		  ACondition = (GetValue(i, clBUDGET) > 30000);
		  break;
	  case 2:
		s = GetValue(i, clPHONE);
		ACondition = (s.Length() > 1) && (s.Pos("8") == 2);
		break;
	  default:
		ACondition = false;
	};
	if (ACondition)
	  ARowIndexList->Add(i);
  };
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::cxDBTreeList1GetScrollbarAnnotationHint(TObject *Sender,
		  TdxScrollbarAnnotationRowIndexLists *AAnnotationRowIndexLists, UnicodeString &AHint)

{
  System::DynamicArray<TdxScrollbarAnnotationKind> AKinds = AAnnotationRowIndexLists->Keys->ToArray();
  TdxScrollbarAnnotationKind AKind;
  int ACount;
  String AAnnotationHint;
  TdxScrollbarAnnotationRowIndexList* AList;
  int ARecordIndex;
  for (int i = 0; i < AKinds.Length; i++)
  {
	AKind = AKinds[i];
	AAnnotationRowIndexLists->TryGetValue(AKind, AList);
	ACount = Min(5, AList->Count);
	for (int j = 0; j < ACount; j++)
	{
	  ARecordIndex = AList->Items[j];
	  switch (AKind)
	  {
		case 0:
			AAnnotationHint = "has vacancy";
			break;
		case 1:
		case dxErrorScrollbarAnnotationID:
		   {
			float f = GetValue(ARecordIndex, clBUDGET);
			AAnnotationHint = "budget is " + FormatFloat("$,0;($,0)", f);
			break;
		   }
		case 2:
			AAnnotationHint = String::Format("phone number is %s", ARRAYOFCONST((GetValue(ARecordIndex, clPHONE))));
			break;
		default:
		    AAnnotationHint = "";
	  };
	  if (AAnnotationHint != "")
	  {
		AAnnotationHint = String::Format("%s %s", ARRAYOFCONST((GetValue(ARecordIndex, cxDBTreeListNAME), AAnnotationHint)));
		AHint = AHint + IfThen(AHint != "", dxCRLF) + AAnnotationHint;
	  };
	};
	if (AList->Count > ACount)
	  AHint = AHint + dxCRLF + "And " + IntToStr(AList->Count - ACount) +  " more records...";
  };
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::cxDBTreeList1KeyDown(TObject *Sender,
		  WORD &Key, TShiftState Shift)
{
  Set <TdxScrollbarAnnotationKind, 0, 255> AKinds;
  AKinds << dxErrorScrollbarAnnotationID;
  if (Key == VK_F4)
	cxDBTreeList1->GoToNextScrollbarAnnotation(AKinds, !Shift.Contains(ssShift));
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::clBUDGETValidateDrawValue(TcxTreeListColumn *Sender,
          TcxTreeListNode *ANode, const Variant &AValue, TcxEditValidateInfo *AData)

{
  if (AValue < 15000)
	AData->ErrorType = eetWarning;
}
//---------------------------------------------------------------------------
void __fastcall TScrollbarAnnotationsDemoMainForm::FormCreate(TObject *Sender)
{
  SCustomAnnotationStrs[0] = "Has open vacancy";
  SCustomAnnotationStrs[1] = "Budget is Greater Than " + CurrToStrF(30000, ffCurrency, 0);
  SCustomAnnotationStrs[2] = "Phone number starts with 8";
  cxDBTreeList1->DataController->FindCriteria->Text = "00000";
  actVacancy->Caption = SCustomAnnotationStrs[0];
  actBudget->Caption = SCustomAnnotationStrs[1];
  actPhoneNumber->Caption = SCustomAnnotationStrs[2];
  cxDBTreeList1->AbsoluteVisibleItems[0]->Expand(false);
}
//---------------------------------------------------------------------------

File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressQuantumTreeList/Demos/Delphi/ScrollbarAnnotationsDemo/ScrollbarAnnotationsDemoMain.pas
unit ScrollbarAnnotationsDemoMain;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics,
  Controls, Forms, Dialogs, DemoBasicMain, Menus, cxClasses, cxLookAndFeels,
  ActnList, ImgList, ComCtrls, StdCtrls, cxGraphics, cxControls, dxCore,
  cxLookAndFeelPainters, cxCustomData, cxStyles, cxTL, cxMaskEdit, cxTextEdit,
  cxCheckBox, cxInplaceContainer, cxDBTL, cxTLData, dxScrollbarAnnotations, cxEdit;

type
  TScrollbarAnnotationsDemoMainForm = class(TDemoBasicMainForm)
    cxDBTreeList1: TcxDBTreeList;
    cxDBTreeListID: TcxDBTreeListColumn;
    cxDBTreeListPARENTID: TcxDBTreeListColumn;
    cxDBTreeListNAME: TcxDBTreeListColumn;
    clBUDGET: TcxDBTreeListColumn;
    clPHONE: TcxDBTreeListColumn;
    cxDBTreeListFAX: TcxDBTreeListColumn;
    cxDBTreeListEMAIL: TcxDBTreeListColumn;
    clVACANCY: TcxDBTreeListColumn;
    ActionList1: TActionList;
    actScrollAnnotationsActive: TAction;
    actVacancy: TAction;
    actBudget: TAction;
    actPhoneNumber: TAction;
    actShowErrors: TAction;
    actShowSearchResults: TAction;
    actShowFocusedRow: TAction;
    actShowSelectedRows: TAction;
    actCustomAnnotationSettings: TAction;
    Scrollannotations1: TMenuItem;
    Active1: TMenuItem;
    N1: TMenuItem;
    Vacancy1: TMenuItem;
    Budgetmorethen300001: TMenuItem;
    Phonenumberstartswith81: TMenuItem;
    Customannotationssettings1: TMenuItem;
    N2: TMenuItem;
    ShowErrors1: TMenuItem;
    ShowSearchResults1: TMenuItem;
    ShowFocusedRow1: TMenuItem;
    ShowSelectedRows1: TMenuItem;
    procedure actCustomAnnotationSettingsExecute(Sender: TObject);
    procedure actScrollAnnotationsActiveExecute(Sender: TObject);
    procedure actShowErrorsExecute(Sender: TObject);
    procedure actShowSearchResultsExecute(Sender: TObject);
    procedure actShowFocusedRowExecute(Sender: TObject);
    procedure actShowSelectedRowsExecute(Sender: TObject);
    procedure actVacancyExecute(Sender: TObject);
    procedure clBUDGETValidateDrawValue(Sender: TcxTreeListColumn; ANode: TcxTreeListNode; const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxDBTreeList1GetScrollbarAnnotationHint(Sender: TObject; AAnnotationRowIndexLists: TdxScrollbarAnnotationRowIndexLists; var AHint: string);
    procedure cxDBTreeList1KeyDown(Sender: TObject; var Key: Word; Shift: TShiftState);
    procedure cxDBTreeList1PopulateCustomScrollbarAnnotationRowIndexList(Sender: TObject; AAnnotationIndex: Integer; ARowIndexList: TdxScrollbarAnnotationRowIndexList);
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  ScrollbarAnnotationsDemoMainForm: TScrollbarAnnotationsDemoMainForm;

implementation

uses
  Math, StrUtils, ScrollbarAnnotationsDemoData, CustomAnnotationSettings;

{$R *.dfm}

procedure TScrollbarAnnotationsDemoMainForm.actCustomAnnotationSettingsExecute(Sender: TObject);
begin
  if frmCustomAnnotationSettings = nil then
  begin
    Application.CreateForm(TfrmCustomAnnotationsettings, frmCustomAnnotationsettings);
    frmCustomAnnotationsettings.Initialize(cxDBTreeList1.ScrollbarAnnotations.CustomAnnotations);
  end;
  frmCustomAnnotationSettings.Show;
end;

procedure TScrollbarAnnotationsDemoMainForm.actScrollAnnotationsActiveExecute(Sender: TObject);
begin
  cxDBTreeList1.ScrollbarAnnotations.Active := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actShowErrorsExecute(Sender: TObject);
begin
  cxDBTreeList1.ScrollbarAnnotations.ShowErrors := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actShowSearchResultsExecute(Sender: TObject);
begin
  cxDBTreeList1.ScrollbarAnnotations.ShowSearchResults := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actShowFocusedRowExecute(Sender: TObject);
begin
  cxDBTreeList1.ScrollbarAnnotations.ShowFocusedRow := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actShowSelectedRowsExecute(Sender: TObject);
begin
  cxDBTreeList1.ScrollbarAnnotations.ShowSelectedRows := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.actVacancyExecute(Sender: TObject);
begin
  cxDBTreeList1.ScrollbarAnnotations.CustomAnnotations[(Sender as TAction).Tag].Visible := (Sender as TAction).Checked;
end;

procedure TScrollbarAnnotationsDemoMainForm.clBUDGETValidateDrawValue(Sender: TcxTreeListColumn; ANode: TcxTreeListNode; const AValue: Variant; AData: TcxEditValidateInfo);
begin
  if AValue < 15000 then
    AData.ErrorType := eetWarning;
end;

procedure TScrollbarAnnotationsDemoMainForm.cxDBTreeList1GetScrollbarAnnotationHint(Sender: TObject; AAnnotationRowIndexLists: TdxScrollbarAnnotationRowIndexLists; var AHint: string);

  function GetValue(ARecordIndex: Integer; AColumn: TcxTreeListColumn): Variant;
  begin
    Result := cxDBTreeList1.AbsoluteVisibleItems[ARecordIndex].Values[AColumn.ItemIndex]
  end;

var
  AKind: TdxScrollbarAnnotationKind;
  AList: TdxScrollbarAnnotationRowIndexList;
  I: Integer;
  ACount: Integer;
  AAnnotationHint, AAdditionalHint: string;
begin
  for AKind in AAnnotationRowIndexLists.Keys do
  begin
    if AKind in (dxAllCustomScrollbarAnnotationKinds + [dxErrorScrollbarAnnotationID]) then
    begin
      AList := AAnnotationRowIndexLists[AKind];
      ACount := Min(5, AList.Count);
      for I := 0 to ACount - 1 do
      begin
        case AKind of
          0:
            AAdditionalHint := 'has vacancy';
          1, dxErrorScrollbarAnnotationID:
            AAdditionalHint := Format('budget is %s', [FormatFloat('$,0;($,0)', GetValue(AList[I], clBUDGET))]);
          2:
            AAdditionalHint := Format('phone number is %s', [GetValue(AList[I], clPHONE)]);
        end;
        AAnnotationHint := Format('%s %s', [GetValue(AList[I], cxDBTreeListNAME), AAdditionalHint]);
        AHint := AHint + IfThen(AHint <> '', dxCRLF) + AAnnotationHint;
      end;
      if AList.Count > ACount then
        AHint := AHint + dxCRLF + Format('And %d more records...', [AList.Count - ACount]);
    end;
  end;
end;

procedure TScrollbarAnnotationsDemoMainForm.cxDBTreeList1KeyDown(Sender: TObject; var Key: Word; Shift: TShiftState);
begin
  if Key = VK_F4 then
    cxDBTreeList1.GoToNextScrollbarAnnotation([dxErrorScrollbarAnnotationID], not (ssShift in Shift))
end;

procedure TScrollbarAnnotationsDemoMainForm.cxDBTreeList1PopulateCustomScrollbarAnnotationRowIndexList(Sender: TObject; AAnnotationIndex: Integer; ARowIndexList: TdxScrollbarAnnotationRowIndexList);

  function GetValue(ARecordIndex: Integer; AColumn: TcxTreeListColumn): Variant;
  begin
    Result := cxDBTreeList1.AbsoluteVisibleItems[ARecordIndex].Values[AColumn.ItemIndex]
  end;

var
  I: Integer;
  ACondition: Boolean;
  AValue: Variant;
  AStr: string;
begin
  for I := 0 to cxDBTreeList1.AbsoluteVisibleCount - 1 do
  begin
    case AAnnotationIndex of
      0:
        begin
          AValue := GetValue(I, clVACANCY);
          ACondition := not VarIsNull(AValue) and AValue;
        end;
      1:
        begin
          ACondition := GetValue(I, clBUDGET) > 30000;
        end;
      2:
        begin
          AValue := GetValue(I, clPHONE);
          if VarIsStr(AValue) then
          begin
            AStr := AValue;
            ACondition := (Length(AStr) > 1) and (AStr[2] = '8');
          end
          else
            ACondition := False;
        end
    else
      ACondition := False;
    end;
    if ACondition then
      ARowIndexList.Add(I);
  end;
end;

procedure TScrollbarAnnotationsDemoMainForm.FormCreate(Sender: TObject);
begin
  inherited;
  SCustomAnnotationstrs[0] := 'Has open vacancy';
  SCustomAnnotationstrs[1] := 'Budget is Greater Than ' + CurrToStrF(30000, ffCurrency, 0);
  SCustomAnnotationstrs[2] := 'Phone number starts with 8';
  cxDBTreeList1.DataController.FindCriteria.Text := '00000';
  actVacancy.Caption := SCustomAnnotationstrs[0];
  actBudget.Caption := SCustomAnnotationstrs[1];
  actPhoneNumber.Caption := SCustomAnnotationstrs[2];
  cxDBTreeList1.AbsoluteVisibleItems[0].Expand(False);
end;

end.
